File: code/ch06-02/stats.erl
%% @author J D Eisenberg <jdavid.eisenberg@gmail.com>
%% @doc Functions for calculating basic statistics on a list of numbers.
%% @copyright 2013 J D Eisenberg
%% @version 0.1

-module(stats).
-export([minimum/1, maximum/1, range/1]).

%% @doc Returns the minimum item in a list of numbers. Fails when given
%% an empty list, as there's nothing reasonable to return.

-spec(minimum(list(number())) -> number()).

minimum(NumberList) ->
  [Result | Rest] = NumberList,
  minimum(Rest, Result).

minimum([], Result) -> Result;

minimum([Head|Tail], Result) ->
  case Head < Result of
    true -> minimum(Tail, Head);
    false -> minimum(Tail, Result)
  end.

%% @doc Returns the maximum item in a list of numbers. Fails when given
%% an empty list, as there's nothing reasonable to return.

-spec(maximum(list(number())) -> number()).

maximum(NumberList) ->
  [Result | Rest] = NumberList,
  maximum(Rest, Result).

maximum([], Result) -> Result;

maximum([Head|Tail], Result) ->
  case Head > Result of
    true -> maximum(Tail, Head);
    false -> maximum(Tail, Result)
  end.

%% @doc Return the range (maximum and minimum) of a list of numbers
%% as a two-element list.
-spec(range([number()]) -> [number()]).

range(NumberList) -> [minimum(NumberList), maximum(NumberList)].

File: code/ch06-02/stats.ex
defmodule Stats do
  @moduledoc """ 
  Functions for calculating basic statistics.
  
  from *Études for Elixir*, O'Reilly Media, Inc., 2013.
  Copyright 2013 by J. David Eisenberg.
  """ 
  @vsn 0.1 

  @doc "Recursively find the minimum entry in a list of numbers."
  @spec minimum([number]) :: number
  
  def minimum(list) do
    [head | tail] = list
    minimum(tail, head)
  end
  
  # When there are no more numbers, return the result.
  
  @spec minimum([number], number) :: number
  
  defp minimum([], result) do
    result
  end
  
  # If the current result is less than the first item in the list,
  # keep it as the result and recursively look at the remainder of the list.
  # Note that you can use a variable assigned as part of a cons in a guard.
  
  defp minimum([head | tail], result) when result < head do
    minimum(tail, result)
  end
  
  # Otherwise, the head of the list becomes the new minimum,
  # and recursively lok at the remainder of the list.
  
  defp minimum([head | tail], _result) do
    minimum(tail, head)
  end

  @doc "Recursively find the maximum entry in a list of numbers."
  @spec maximum([number]) :: number
  
  def maximum(list) do
    [head | tail] = list
    maximum(tail, head)
  end
  
  # When there are no more numbers, return the result.
  
  @spec maximum([number], number) :: number
  
  defp maximum([], result) do
    result
  end
  
  # If the current result is greater than the first item in the list,
  # keep it as the result and recursively look at the remainder of the list.
  
  defp maximum([head | tail], result) when result > head do
    maximum(tail, result)
  end
  
  # Otherwise, the head of the list becomes the new maximum,
  # and recursively look at the remainder of the list.
  
  defp maximum([head | tail], _result) do
    maximum(tail, head)
  end
  
  # @doc "Find the range of a list of numbers as a list [min, max]."
  @spec range([number]) :: [number]
  
  def range(list) do
    [minimum(list), maximum(list)]
  end
end
